### DISCUSSION TITLE
TickerHandler - What did I do?

### DISCUSSION THREAD

**Discussion by RandomEC** [evennia/evennia#2721] [Evennia Questions]
So, I had a few lingering tickers that were created before I really knew how to use them, that weren't getting rid of themselves appropriately. As a result, I created the following script to clean them up:
```
from evennia import TICKER_HANDLER as tickerhandler

class TickerCleanupFixAreaNames(DefaultScript):

    # Start with scripts/start scripts.FixAreaNames

    def at_script_creation(self):
        self.key = "ticker_cleanup_script"
        self.desc = "Cleans up old tickers"
        self.persistent = True

        tickerhandler.clear(interval=900)
        tickerhandler.clear(interval=1800)
```

After running it, I got a large error, which I foolishly did not preserve. Fortunately, however, anytime I try to start combat (which starts a ticker), I get the following error, which looks virtually identical to the original error. I also cannot reload or restart within the mud, nor stop Evennia from the command prompt. What did I do wrong? How do I fix it?

The only thing I can think is that the flat API says the following about .clear: "This is the only supported way to kill tickers related to non-db objects." I did use it on db objects, does the foregoing also mean that it should not be used on db objects?

```
Traceback (most recent call last): 
File "c:\users\bradm\muddev\evennia\evennia\commands\cmdhandler.py", line 644, in _run_command 
ret = cmd. func() File ".\commands combat_commands.py", line 464, in func 
combat = rules_combat.create_combat(caller, target) File "C:\Users\bradm\muddev\echoes\world\rules_combat.py", line 25, in create_combat 
combat = create_object("commands.combat_commands.Combat", key=("combat_handler_%S" % attacker.location.db. vnum)) 
File "c:\users\bradm\muddev\evennia\evennia\utils\create.py", line 171, in create_object 
new_object. save() File "c:\users\bradm\muddev\evennia\evennia\utils\idmapper\models.py", line 442, in save 
super().save(*args, **kwargs) File "C:\Users\bradm\muddev\evenv\Lib\site-packages\django\db\models\base.py", line 740, in save 
force_update=force_update, update_fields=update_fields) File "C:\Users\bradm\muddev\evenv\Lib\site-packages\django\db\models\base.py", line 789, in save_base 
update_fields=update_fields, raw=raw, using=using, File "C:\Users\bradm\muddev\evenv\Lib\site-packages\django\dispatch\dispatcher.py", line 182, in send 
for receiver in self._live_receivers ( sender) File "C:\Users\bradm\muddev\evenv\Lib\site-packages\django\dispatch\dispatcher.py", line 182, in <listcomp> 
for receiver in self._live_receivers ( sender) File "c:\users\bradm\muddevlevennialevennia\typeclasses\models.py", line 70, in call_at_first_save 
instance.at_first_save(). File "c:\users\bradm\muddevlevennia\evennia\objects\objects.py", line 1141, in at_first_save 
self.at_object_creation() File ".\commands\combat_commands.py", line 22, in at_object_creation 
tickerhandler.add(2, self.at_repeat) File "c:\users\bradm\muddev\evennialevennia\scripts\tickerhandler.py", line 533, in add 
self.save(). File "c:\users\bradm\muddev\evennia\evennia\scripts\tickerhandler.py", line 416, in save 
for store_key, (args, kwargs) in self.ticker_storage. items (). File "c:\users\bradm\muddev\evennialevennia\scripts\tickerhandler.py", line 416, in <dictcomp> 
for store_key, (args, kwargs) in self.ticker_storage. items () ValueError: too many values to unpack (expected 2) 
An untrapped error occurred. 
(Traceback was logged 22-04-08 08:39:23-04). 
```

As usual, thanks in advance for any help anyone can give.

Edit: I did chase back the final reference, in which the documentation for "save" says that it saves ticker_storage as a serialized string into a temporary ServerConf field. If it's temporary, maybe just restarting the whole computer will get rid of it? I'll try that when I go home. The comment above line 416 says "remove any subscriptions that lost its object in the interim", which leads me to believe that maybe my suspicion above that I shouldn't have used clear on tickers for objects that still exist was right?

**Comment by kovitikus** [comment]
Based on the tracelog you pasted it seems like you are executing a command that creates a `CombatHandler` typeclassed object using the `Combat` class located in the `combat_commands` module.

In that `Combat` class you have an `at_object_creation()` method where you are placing your ticker creation, `tickerhandler.add(2, self.at_repeat)`

At first glance it seems like the tickerhandler is yelling at you that you gave it too much information, but you didn't.

The problem seems to actually be happening when the tickerhandler attempts to save that new ticker and the list of saved tickers it was working with doesn't have the same number of entries it expected.

Somewhere you are deleting them in a way that the tickerhandler isn't staying in sync with your changes.

This is my best guess. Others will have to elaborate or correct me.

**Reply by RandomEC** [reply]
Kovitikus -

Since I have been playing a lot, and have started tons of combats prior to this, my guess is that the script I ran is the cause, and that the error is a hangover from that, and combat trying to start is just what's getting it to rear its head.

Thanks!

**Reply by kovitikus** [reply]
https://github.com/evennia/evennia/blob/master/evennia/scripts/tickerhandler.py#L413

Ok, based on the line 416 in the tickerhandler `save()` method, this comment here states this section of the code is for...

```py
            # remove any subscriptions that lost its object in the interim
            to_save = {
                store_key: (args, kwargs)
                for store_key, (args, kwargs) in self.ticker_storage.items()
                if (
                    (
                        store_key[1]
                        and ("_obj" in kwargs and kwargs["_obj"].pk)
                        and hasattr(kwargs["_obj"], store_key[1])
                    )
                    or store_key[2]  # a valid method with existing obj
                )
            }  # a path given
```

It looks like the purpose here is to create a dictionary called `to_save`, which from what I can gather does some sort of filtration to create a final dictionary that has tickers still alive?

In any case, I still think it stands that you are messing with the tickerhandler's saved ticker data in a way that thetickerhandler isn't expecting it.

**Reply by RandomEC** [reply]
kovitikus -

Yeah, I think that is right. My suspicion is that I shouldn't have used clear because it is for non-db object tickers, so when I used it, it cleared the ticker, but didn't decrement something the way it would if it was for a db object ticker.

So, I just need to know how to fix it. Hopefully without completely redoing my database. ;)

Thanks again.

**Comment by Griatch** [comment]
@RandomEC This has been sitting for a while, did it work out for you in the end?
